statemachine/statemachine.py

```diff
@@ -302,25 +302,38 @@ def _activate(self, event_data: EventData):
         source = event_data.state
         target = transition.target
 
-        result = transition.before.call(*event_data.args, **event_data.extended_kwargs)
+        result_list = transition.before.call(
+            *event_data.args, **event_data.extended_kwargs
+        )
+
         if source is not None and not transition.internal:
-            source.exit.call(*event_data.args, **event_data.extended_kwargs)
+            result_list += source.exit.call(
+                *event_data.args, **event_data.extended_kwargs
+            )
 
-        result += transition.on.call(*event_data.args, **event_data.extended_kwargs)
+        result_list += transition.on.call(
+            *event_data.args, **event_data.extended_kwargs
+        )
 
         self.current_state = target
         event_data.state = target
 
         if not transition.internal:
-            target.enter.call(*event_data.args, **event_data.extended_kwargs)
-        transition.after.call(*event_data.args, **event_data.extended_kwargs)
+            result_list += target.enter.call(
+                *event_data.args, **event_data.extended_kwargs
+            )
+
+        result_list += transition.after.call(
+            *event_data.args, **event_data.extended_kwargs
+        )
 
-        if len(result) == 0:
-            result = None
-        elif len(result) == 1:
-            result = result[0]
+        result_list = [r for r in result_list if r not in [None, []]]
```

Comment:
**suggestion:** Add a const at the top of the file like `EMPTY_RETURN_VALUES`, the values should be a `set` for _maybe_ faster `in` checks (`O(1)` instead of `O(n)`).

```py
EMPTY_RETURN_VALUES = {None, []}
```

```suggestion
        result_list = [r for r in result_list if r not in EMPTY_RETURN_VALUES]
```

Reply:
Thank you for the suggestion!

I've benchmarked it and you are right.

Before:
```sh
% python -m timeit -s 'result_list = [None, "foo", None, "bar", None, "baz"]' '[r for r in result_list if r not in [None, []]]'
500000 loops, best of 5: 898 nsec per loop
```

Module scoped list:
```sh
% python -m timeit -s 'EMPTY_RETURN_VALUES = [None, []]; result_list = [None, "foo", None, "bar", None, "baz"]' '[r for r in result_list if r not in EMPTY_RETURN_VALUES]'
500000 loops, best of 5: 601 nsec per loop
```

Module scoped tuple:
```sh
% python -m timeit -s 'EMPTY_RETURN_VALUES = (None, []); result_list = [None, "foo", None, "bar", None, "baz"]' '[r for r in result_list if r not in EMPTY_RETURN_VALUES]'
500000 loops, best of 5: 590 nsec per loop
```

A set is not applicable here, because a list is not a hashable type :crying_cat_face: